Blog for Search Improvements (#47)
rtfd/blog: search-improvements.rst

.. post:: August 14, 2018
   :tags: gsoc, search, feature
   :author: Safwan
   :location: DHA

Improved Search
==================================
Have you ever used the search functionality inside documentations hosted by Read The Docs?
Mostly yes. Search can always be improved and in the era of Search Engines, the
documentation search also plays a big role for the users. We have realized the importance,

> and got our GSoC student interested to take the challenge. The main goal was to refactor the

Reviewer: I think it's ok to write this in first person (I, me, etc.). Introduce yourself! You are doing great work and you get to take credit for it.

Reviewer: I see that you introduce yourself below. That's good but I still think you can write in first person.

search code together with upgrading the backend Search engine, as well as adding more features
to the search functionality like exact match search, case insensitive search,
search as you type, suggestions etc.

Background
^^^^^^^^^^
Search is a vital part of any documentation hosting platform as people can get the
information they need. As a documentation hosting platform, same rule applies for
Read The Docs. Because of having a small core team, the search functionality
of Read The Docs has lagged behind for quite a while now. Initially the search code
was voluntarily contributed by `Rob Hudson`_,  `back in 2013`_ and then improved by other
contributors. The time span and continuous technology growing made the search
module got older and it had become outdated. Moreover The search infrastructure was already
vulnerable as we were using `Elasticsearch 1.3.x`_ which was already reached its
End of Life in 2016. Therefore, the upgradation of search infrastructure was badly needed.

> Google Summer of Code

Reviewer: This is an opinion, so take it for what it's worth, but I'd put this section above "Background". I think it flows in a logical order after the introduction.

~~~~~~~~~~~~~~~~~~~~~
Google Summer of Code is a global program where students work with an open source organization
on a 3 month programming project. The core team of Read The Docs proposed some `Project Ideas`_,
one of them was **Refactor & improve our search code**. Our contributor (me) `Safwan Rahman`_,
an undergraduate university student and open source contributor from Bangladesh was keen
to get his hand dirty with Elasticsearch_ and grasped the opportunity to do so by applying
for this project and got accepted.

> Safwan_ has worked full time for last 3 months to upgrade the whole codebase

Reviewer: I'd use bullets for this section. It will come across better than a big block of text links.

to compatable with `Elasticsearch 6.x`_ and also implemented various features
like `Exact Matching Search`_, `Case Insensitive Search`_, `Improved Result Order`_,
`Zero Downtime Indexing`_ and many more. Together with this, he is willing to implement
`Search as You Type and Autocomplete`_, `Code Search`_ and many more.  All of his search related work can be
looked in the `Search Project Board`_.

> 

Reviewer: Should we link to past GSOCs for Read the Docs? I actually looked and didn't see blogs for those. This is more of a question for @ericholscher.

Reviewer 2: This is the first GSOC we've done.

Sphinx/MkDocs Built in Search vs Read The Docs Search
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Sphinx/MkDocs already have built in search functionality. But the features are very limited and heavy.
At Read The Docs, we have felt the limitations and therefore we index the documentations in our
Elasticsearch_ index so that we can provide better results to the users. Currently, we do not index

> MkDocs documents to Elasticsearch_, but `any kind of help is welcome`_.

Reviewer: I think we can expand a little bit on this section. Just saying it is heavy doesn't say much. Here's a few brief ideas worth mentioning:

- searching across different projects
- more advanced query syntax

- `Improved Result order`_: The result order is improved dramatically. Therefore if you Search
  with `Foo Bar`, you will first see the documentations which have both `Foo Bar`, then
  you will see other documentations which have either `Foo` or `Bar`

- `Auto Removing from Search Index`_: In the past, if any page got removed from documentations,
  the page was still available in documentations. But from now on, the page will be removed
  automatically from the search index as soon as its no longer in the documentation.

- `Zero Downtime Indexing`_: As search is a important part of documentation, there will be no
  downtime while we reindex our Elasticsearch Index. So the search will be much more reliable
  than before.

Performance Improvements
^^^^^^^^^^^^^^^^^^^^^^^^
Performance is always improtant for search. Because of our search was fast enough,
we did not focus to improve the performance. As we have rewritten all the codes with
new search backend engine, we got 4x improved performance as well.

There are 4 kind of search functionality in Read The Docs. The improvements are as following:

> - `File Search`_

Reviewer: I'd give a one sentence blurb explaining what each of "File Search", "Projects Search", are. Those two are somewhat obvious but "Doc Search" and "Project Doc Search" are not.

  - Before: 1461 ms
  - After: 200 ms

- `Projects Search`_

  - Before: 455 ms
  - After: 162 ms

- `Doc Search API`_

  - Before: 746 ms
  - After: 173 ms

- `Project Doc Search`_

  - Before: 750 ms
  - After: 270 ms

Code Improvements
^^^^^^^^^^^^^^^^^
Code quality is very much important in development world, specially in open source.
As we have rewritten the search functionality from scratch, the code quality

> is improved in many ways. Please let us know if you are interested to contribute.

Reviewer: This section ends a little abruptly. Do you see any next steps after GSOC where this can be improved? Anything you didn't get to that you think would be a good idea?

> .. post:: August 14, 2018

Reviewer 2: Just need to update this :)

Author: @ericholscher updated!

.. post:: August 14, 2018
   :tags: gsoc, search, feature
   :author: Safwan
   :location: DHA

Improved Search
==================================
Have you ever struggled with a poorly documented software project?
What about a well documented project but you can't find the right section inside the docs?
The Read the Docs `core team`_ has realized the importance of good search for documentation
and got me to take the challenge as a Google Summer of Code student.
The main goal of my GSoC project was to refactor the search code together with upgrading the backend
search engine, as well as adding more features to our search functionality like exact match search,
case insensitive search, search as you type, suggestions and more.

Google Summer of Code

> ^^^^^^^^^^^^^^^^^^^^^ Contributors Wanted

Reviewer: There's something wrong with the heading here.

![screen shot 2019-01-28 at 11 31 36 am](https://user-images.githubusercontent.com/185043/51861159-57ad9580-22f0-11e9-811f-92e2f3c8d554.png)

```suggestion
^^^^^^^^^^^^^^^^^^^^^
```

Google Summer of Code is a global program where students work with an open source organization
on a 3 month programming project. The core team of Read The Docs proposed some `Project Ideas`_,

> one of them was **Refactor & improve our search code**. I (`Safwan Rahman`_) was keen to get my hand dirty with

Reviewer 2: `hands dirty`

Elasticsearch_ and grasped the opportunity to do so by applying
for this project and I got accepted.

> I have worked full time for last 3 months to upgrade the whole codebase

Reviewer 2: ```suggestion
I worked full time from August to September 2018 to upgrade the whole codebase
```

to compatable with `Elasticsearch 6.x`_ and also implemented various features

> as like:

Reviewer 2: remove `as`

- `Exact Matching Search`_
- `Case Insensitive Search`_

> - `Improved Result Order`_,

Reviewer 2: ```suggestion
- `Improved Result Order`_
```

- `Zero Downtime Indexing`_

> Together with this, I am willing to implement more features like following:

Reviewer: like following -> like the following

> Together with this, I am willing to implement more features like the following:

Reviewer 2: ```suggestion
Together with this, we are planning more features like the following:
```

Together with this, I am willing to implement more features like following:

- `Search as You Type and Autocomplete`_
- `Code Search`_ 

> All of my search related work can be looked in the `Search Project Board`_.

Reviewer 2: `All of my search related work can be seen in the`

Background
^^^^^^^^^^
Search is a vital part of any documentation hosting platform, so people can get the
information they need. As a documentation hosting platform, the same rules apply to
Read The Docs. Because of having a small core team, the search functionality
of Read The Docs has lagged behind for quite a while now. Initially the search code
was voluntarily contributed by `Rob Hudson`_,  `back in 2013`_ and then improved by other
contributors. The search infrastructure was already outdated as
Read The Docs were using `Elasticsearch 1.3.x`_ which was already reached its

> End of Life in 2016. Therefore, the upgradation of search infrastructure was badly needed.

Reviewer: Therefore, the upgradation of search infrastructure -> Upgrading the search infrastructure

Built in Search vs Read The Docs Search
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Both Sphinx and MkDocs already have built in search functionality. But the features are very limited.
At Read The Docs, we have felt the limitations and therefore we index the documentations in our
Elasticsearch_ index so that we can provide better search experience like:

- Search across multiple projects
- Advanced query syntax
- Search inside subprojects
- Improved search result order
- Public Search API (Documentation pending)

**Currently, we do not index MkDocs documents with** Elasticsearch_,
but `any kind of help is welcome`_.

New Features
^^^^^^^^^^^^

> In the 3 months of full time work, I have implemented couple of features including

Reviewer: implemented couple -> implemented a couple

> In the 3 months of full time work, I have implemented a couple of features including

Reviewer 2: ```suggestion
In the 3 months of full time work, I have implemented these features along with 
```

many bug fixes. Some of the major features are as following:

- `Exact Matching Search`_: Exact matching is one of our most highly requested

>   features for Read the Docs. Now you can search for exact word in documentations

Reviewer: documentations -> documentation

  by having your query inside quotation (`""`). So if you search

>   for `"Here is foo"` (with the quotation), you will get all the documentations where the full

Reviewer: documentations -> documentation

>   `Here is foo` phrase exist.

Reviewer: exist -> exists.

Single backticks don't really do anything on the stylesheet we are using. If you build the blog, you'll see this:

<img width="550" alt="screen shot 2018-10-04 at 1 25 00 pm" src="https://user-images.githubusercontent.com/185043/46500952-f0b42000-c7d8-11e8-8516-e604fd0de6f2.png">

- Simple Query String Syntax support: Now you can use `Simple Query String Syntax`_ for

>   searching. For example, if you search with `Mozilla +-Firefox`, you will get documentations

Reviewer: documentations -> documentation

  where the `Mozilla` word is present, but `Firefox` word is not present.

>   For more information, you can look at `Simple Query String Syntax`_ documentation.

Reviewer: look at -> look at the

- `Case Insensitive Search`_: The search is now case insensitive. So if you search for `Foo`,

>   you will get all the documentations which have one of either `Foo`, `FOO`, `foo` or `fOO`.

Reviewer: documentations -> documentation

- `Improved Result order`_: The result order is improved dramatically. Therefore if you Search
  with `Foo Bar`, you will first see the documentations which have both `Foo Bar`, then

>   you will see other documentations which have either `Foo` or `Bar`

Reviewer: documentations -> documentation

- `Auto Removing from Search Index`_: In the past, if any page got removed from documentations,

>   the page was still available in documentations. But from now on, the page will be removed

Reviewer: documentations -> documentation

Reviewer: If you want to mix up your verbiage a bit, you can use "docs" but "documentations" (while technically a word) is extremely rare and probably shouldn't be used.

  automatically from the search index as soon as its no longer in the documentation.

> - `Zero Downtime Indexing`_: As search is a important part of documentation, there will be no

Reviewer: a important -> an important

  downtime while we reindex our Elasticsearch Index. So the search will be much more reliable
  than before.

Performance Improvements
^^^^^^^^^^^^^^^^^^^^^^^^
Performance is always important for search. Because our search was fast enough,

> we did not focus to improve the performance. As we have rewritten all the codes with

Reviewer: codes -> code

Performance is always improtant for search. Because of our search was fast enough,
we did not focus to improve the performance. As we have rewritten all the codes with

> new search backend engine, we got 4x improved performance as well.

Author: Mostly newer elasticsearch as well as new optimized algorithms

There are 4 kind of search functionality in Read The Docs. The improvements are as following:

> - `File Search`_: Search accross all the projects of Read The Docs.

Reviewer 2: `across` -- you should run English spell check on this whole document and fix things with it.

  - After: 200 ms

- `Projects Search`_: Search for projects hosted in Read The Docs.

  - Before: 455 ms
  - After: 162 ms

- `Doc Search API`_: The Public API which is used to perform search in documentation.

  - Before: 746 ms
  - After: 173 ms

- `Project Doc Search`_: Search inside a project documentations from readthedocs.org domain.

  - Before: 750 ms
  - After: 270 ms

Code Improvements
^^^^^^^^^^^^^^^^^

> Code quality is very much important in development world, specially in open source.

Reviewer: very much important -> very important

> 

Reviewer 2: We should show the code that we used to establish these benchmarks, so it's reproducable.

Author: I do not have any code ready for this. its possible to check the browser network bar to get the timing.

Reviewer 2: Is this still the case in prod? I don't want to publish benchmarks without having them be reproducible in some fashion.

Reviewer 2: We should remove this section unless we can publish some kind of benchmark, at least for the current search. A simple `time curl https://readthedocs.org/search/` would be fine.

As I have rewritten the search functionality from scratch, the code quality

> is improved in many ways like test coverage and documentations. So its easy for

Reviewer: documentations -> documentation

any contributor to start working on the search functionality

Contributors Wanted

> ^^^^^^^^^^^^^^^^^^

Reviewer: You need one more `^` on this line:

> search-improvements.rst:135: WARNING: Title underline too short.

As Read The Docs is an open source project backed by a small team of developers,
most of them are busy to keep things up and running only. Therefore, its quite
hard for them to take time to implement new features. If you know some bit of
Django or Python and Elasticsearch, you can contribute into the search functionality
of Read The Docs. If you need any support to start contributing, you can get in touch with
me or any member of  Read The Docs team. You can find all of us at `#readthedocs` freenode
IRC channel or `readthedocs gitter`_ channel. I am `safwan` at IRC and `@safwanrahman`
at gitter.

Conclusion
^^^^^^^^^^

> To conclude, I must say that the Search improvement in Read The Docs was very much 

Reviewer: very much necessary -> very necessary

> necessary and I could improve it in some extent in short amount of time. 

Reviewer: I think there's a couple words missing in this sentence:

    "... and I could improve it in some extent in short amount of time." ->
    and I'm glad I could improve it in such a short amount of time

```suggestion
necessary and I'm glad I could improve it in such a short amount of time. 
```

> ^^^^^^^^^^

Reviewer: One thing missing from the conclusion that I think should be in there is about how much you learned as part of GSoC?

Author: I can include that, but I think this blog post is more focused on the search improvement of read the docs. so I am not sure if it should be included

To conclude, I must say that the Search improvement in Read The Docs was very much 
necessary and I could improve it in some extent in short amount of time. 

> There can be infinity number of ways it can be improved and I believe we can compete

Reviewer: There can be infinity number -> There are an infinite number

with major search engines in terms of documentation searching.

> As I have worked for only 3 months full time, some compelling features are left behind

Reviewer: This sentence is a bit awkward and maybe it should be reworded along the lines of:

    Due to the constraints of only working for three months,
    a number of compelling features were left out such as...

without implementing like `Search as You Type and Autocomplete`_,
`Code Search`_ functionality. Moreover, proper documentation is needed for the search
architecture. I have tried to write test cases for most of the scenario, but because of

> time constrains, a lot of code is out of test coverage. I strongly hope that

Reviewer: I would start a new paragraph before "I strongly"

I strongly hope that we will get the left behind work done within a short amount of time.
This can be done easily if we get more contributors donate their time for improving Read The Docs.

> We dont need superhero or coding guru, just need people who understand Python, Django and

Reviewer: We don't need **a** superhero....

```suggestion
We don't need a superhero or coding guru, you just need people who understand Python, Django and
```

> Elasticsearch and have some time to write some code for us. You are **Superhero** to us

Reviewer: You are superhero -> You are a superhero

> Elasticsearch and have some time to write some code for us. You are **SuperHero** to us

Reviewer: SuperHero -> a Superhero

Reviewer: Python/Django and Elasticsearch and -> Python, Django, Elasticsearch, and

if you can lend your time and effort to improve Read The Docs.

.. _Rob Hudson: https://github.com/robhudson
.. _back in 2013: https://github.com/rtfd/readthedocs.org/pull/493
.. _Elasticsearch: https://www.elastic.co/products/elasticsearch
.. _Elasticsearch 1.3.x: https://www.elastic.co/guide/en/elasticsearch/reference/1.3/index.html
.. _Elasticsearch 5.x: https://www.elastic.co/guide/en/elasticsearch/reference/5.4/index.html
.. _Elasticsearch 6.x: https://www.elastic.co/guide/en/elasticsearch/reference/6.3/index.html
.. _Elasticsearch 6.x has major changes: https://www.elastic.co/guide/en/elasticsearch/reference/current/release-notes-6.0.0.html

> .. _Project Ideas: https://git.io/fN9GK

Reviewer 2: Why is this a short url?

Author: Because the URL is so big as I have taken it from github commit tag!
so prefered to use github short url

.. post:: August 14, 2018
   :tags: gsoc, search, feature

>    :author: Safwan

Reviewer: For authors, you should have an entry here: https://github.com/rtfd/blog/blob/master/conf.py#L47

>    :location: DHA

Reviewer: To make the location work, you need an entry here: https://github.com/rtfd/blog/blob/master/conf.py#L54

Search Improvements
====================

> 

Reviewer: A very short paragraph here explaining the biggest goals of the search upgrade would be good, I think. You want to get people interested so they read further.
